### PR TITLE
[Instrumentation.StackExchangeRedis] Connection registration changes

### PR DESCRIPTION
**Pull request by lachmatt**
Addresses https://github.com/open-telemetry/opentelemetry-dotnet/issues/1900#issue-829002183

## Changes

Add an ability to track connections created after tracer is initialized.

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
~~* [ ] Design discussion issue #~~
~~* [ ] Changes in public API reviewed~~

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/IConnectionRegistry.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+/// <summary>
+/// Registers connection for instrumentation purposes.
+/// </summary>
+public interface IConnectionRegistry
+{
+    /// <summary>
+    /// Registers connection with current instrumentation instance.
+    /// </summary>
+    /// <param name="connectionMultiplexer">Connection to be tracked by instrumentation.</param>
+    void Register(IConnectionMultiplexer connectionMultiplexer);
```
Comment (pellared) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/IConnectionRegistry.cs:30]: I am not sure about the naming. Maybe something like `Monitor`, `Instrument`, `Track` would be more meaningful?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 2]
```diff
@@ -68,22 +75,78 @@ public static class TracerProviderBuilderExtensions
...
         });
     }
 
     private static TracerProviderBuilder AddRedisInstrumentation(
         TracerProviderBuilder builder,
-        IConnectionMultiplexer connection,
         StackExchangeRedisCallsInstrumentationOptions options,
-        Action<StackExchangeRedisCallsInstrumentationOptions> configure)
+        Action<StackExchangeRedisCallsInstrumentationOptions> configure,
+        out Action<IConnectionMultiplexer> registration,
```
Comment (pellared) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:131]: How about

```suggestion
        out IConnectionRegistry registry,
```

? 

Personally I find the initialization of the registry using `out defaultRegistry.RegistrationCallback` too smart 😉

**Comment by Kielek** [comment]
There is no package owner for StackExchangeRedis.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs (hunk 2 above)
Comment (lachmatt) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs]: changed in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1139/commits/d688db626c37454828d709eff9a2f8a2a93ed542

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 3]
```diff
@@ -26,6 +26,13 @@ namespace OpenTelemetry.Trace;
 /// </summary>
 public static class TracerProviderBuilderExtensions
 {
+    /// <summary>
+    /// Enables automatic data collection of outgoing requests to Redis.
+    /// </summary>
+    /// <param name="builder"><see cref="TracerProviderBuilder"/> being configured.</param>
+    /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
+    public static TracerProviderBuilder AddRedisInstrumentation(this TracerProviderBuilder builder) => AddRedisInstrumentation(builder, null, null);
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs:34]: Please add names to null parameters.

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs [hunk 4]
```diff
@@ -68,22 +75,77 @@ public static class TracerProviderBuilderExtensions
...
+
+    /// <summary>
+    /// Enables automatic data collection of outgoing requests to Redis.
+    /// </summary>
+    /// <param name="builder"><see cref="TracerProviderBuilder"/> being configured.</param>
+    /// <param name="connectionRegistry">Registry to register connections with.</param>
+    /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
+    public static TracerProviderBuilder AddRedisInstrumentation(
+        this TracerProviderBuilder builder,
+        out IConnectionRegistry connectionRegistry) => AddRedisInstrumentation(builder, out connectionRegistry, null);
```
[the remark this answers is not in the document]
Comment (lachmatt) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs]: fixed in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1139/commits/04e3740a95a918935c344add1505abfe68ad240b

**Review by pellared — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/IConnectionRegistry.cs (hunk 1 above)
Comment (pellared) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/IConnectionRegistry.cs]: Addressed

Comment (lachmatt) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/IConnectionRegistry.cs]: addressed in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1139/commits/10ef066e34334b57badd1dee5af6af7cce2e3ee0 and https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1139/commits/da57a4733e8e6b317fc74575d4eebfa959d9775c

**Review by pjanotti — COMMENTED** [review]
The main question is why is necessary to return an `instrumenter`, since it still has usability problems. Ideally one should be able to just use an extension method on the connection, something like `connection.WithOpenTelemetryInstrumentation()`.

File: examples/redis/Examples.StackExchangeRedis/Program.cs [hunk 5]
```diff
@@ -28,19 +28,21 @@ public static void Main()
...
          */
 
-        // connect to the redis server. The default port 6379 will be used.
-        var connection = ConnectionMultiplexer.Connect("localhost");
-
         // Configure exporter to export traces to Zipkin
         using var tracerProvider = Sdk.CreateTracerProviderBuilder()
                 .AddConsoleExporter()
-                .AddRedisInstrumentation(connection, options =>
+                .AddRedisInstrumentation(out var instrumenter, options =>
```
[the remark this answers is not in the document]
Comment (lachmatt) [reply examples/redis/Examples.StackExchangeRedis/Program.cs:34]: I think this is a minimal solution to the problem linked in description.

**Review by CodeBlanch — CHANGES_REQUESTED** [review]
@lachmatt

Hey it looks like we already have [IServiceProvider](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/6b228e0a02e7752a327c100ef630a6138e4089dd/src/OpenTelemetry.Instrumentation.StackExchangeRedis/TracerProviderBuilderExtensions.cs#L62) support in OpenTelemetry.Instrumentation.StackExchangeRedis. Couldn't you achieve essentially what this PR is seeking to enable by using `IServiceCollection` \ `IServiceProvider`?

Manual creation

```csharp
using var provider = Sdk.CreateTracerProviderBuilder()
   .ConfigureServices(services => services.AddSingleton<IConnectionMultiplexer>(sp => RedisHelper.InitializeConnection(sp)))
   .AddRedisInstrumentation()
   .Build();
```

Hosting creation

```csharp
appBuilder.Services.AddSingleton<IConnectionMultiplexer>(sp => RedisHelper.InitializeConnection(sp));
appBuilder.Services.AddOpenTelemetry()
   .WithTracing(builder => builder.AddRedisInstrumentation());
```

**Comment by lachmatt** [comment]
@CodeBlanch 
Issue linked in the description mentions a scenario when connection is not available at startup, but is created on (first) demand later in an app lifecycle. Changes from this PR should help with that.
If I'm not mistaken, what you described changes how connection is passed (through container instead of being passed directly to the extension method), but still requires connection to be created at startup (when instrumentation is initialized).

**Comment by CodeBlanch** [comment]
@lachmatt It is really up to the app. In that snippet I sent above I'm registering a factory for the connection into the service collection. So the connection won't actually be started until that factory is executed (when it is first requested from the `IServiceProvider`).

**Comment by pjanotti** [comment]
@CodeBlanch let me check if I understand your suggestion. You propose adding a static method, `RedisHelper.InitializeConnection`, to `OpenTelemetry.Instrumentation.StackExchangeRedis`, correct? This method will take care of "instrumenting" the connection. If that is the case I agree that it will make the usability of the package much better: it removes the need to pass the connection and also the out parameter currently being added in this PR.

In retrospect, it is not a good practice to pass instances created using the types from the targeted library in the extension methods for the OTel builders, see https://github.com/open-telemetry/opentelemetry-dotnet/issues/1900 and https://github.com/open-telemetry/opentelemetry-dotnet/pull/4347#discussion_r1160092578

A step further will be to mark the `AddInstrumentation` overload receiving the connection as obsolete, per [opentelemetry-dotnet/VERSIONING.md](
https://github.com/open-telemetry/opentelemetry-dotnet/blob/586aa7eb6d33a6a975b4ec78fdf1cf4d6b93c184/VERSIONING.md?plain=1#L14-L18).

**Comment by CodeBlanch** [comment]
> let me check if I understand your suggestion. You propose adding a static method, RedisHelper.InitializeConnection, to OpenTelemetry.Instrumentation.StackExchangeRedis, correct?

Not exactly. What I'm saying is this code is already hooked up to check the `IServiceProvider` for the `IConnectionMultiplexer` so if users don't want to supply the instance when they call `AddRedisInstrumentation` they can use the `IServiceCollection` to manage it.

Let's take the example from the README showing the new API in use:

```csharp
        using var tracerProvider = Sdk.CreateTracerProviderBuilder()
            .AddRedisInstrumentation(out var connectionInstrumenter)
            .AddConsoleExporter()
            .Build();

        // Connect to the server.
        using var connection = ConnectionMultiplexer.Connect("localhost:6379");
        connectionInstrumenter.Instrument(connection);
```

Today you could do this to accomplish the same:

```csharp
        using var tracerProvider = Sdk.CreateTracerProviderBuilder()
            .ConfigureServices(services => services.AddSingleton(sp => ConnectionMultiplexer.Connect("localhost:6379")))
            .AddRedisInstrumentation()
            .AddConsoleExporter()
            .Build();
```

A more useful real-world example would probably be more like...

```csharp
appBuilder.Services.AddSingleton(sp => {
   var myRedisOptions = sp.GetRequiredService<IOptions<RedisOptions>>().Value;
   return ConnectionMultiplexer.Connect(myRedisOptions.Endpoint);
});

appBuilder.Services.AddOpenTelemetry().WithTracing(builder => builder
    .AddRedisInstrumentation()
    .AddConsoleExporter());
```

**Comment by pjanotti** [comment]
Thanks, now I see what you are saying. This should satisfy many (most?) use cases since typical server applications are expected to be using an `IServiceProvider`. Let's hear what people affected by [#1900](https://github.com/open-telemetry/opentelemetry-dotnet/issues/1900) think: @tbolon @philipbawn @adamhathcock @lucasoares @Norbe3t, please, let's know if the usage of `IServiceProvider` as suggested above fixes the issue for your case.

From my perspective, it still forces the call to `ConnectionMultiplexer.Connect` to happen in specific places. What I really would like is to decouple the connection instrumentation from the tracer builder extension method. Perhaps, not as important for StackExchangeRedis which typically uses a singleton for the connection object but developers may see it as a template and repeat the pattern.

**Comment by lucasoares** [comment]
> Thanks, now I see what you are saying. This should satisfy many (most?) use cases since typical server applications are expected to be using an `IServiceProvider`. Let's hear what people affected by [#1900](https://github.com/open-telemetry/opentelemetry-dotnet/issues/1900) think: @tbolon @philipbawn @adamhathcock @lucasoares @Norbe3t, please, let's know if the usage of `IServiceProvider` as suggested above fixes the issue for your case.
> 
> From my perspective, it still forces the call to `ConnectionMultiplexer.Connect` to happen in specific places. What I really would like is to decouple the connection instrumentation from the tracer builder extension method. Perhaps, not as important for StackExchangeRedis which typically uses a singleton for the connection object but developers may see it as a template and repeat the pattern.

Yeah, for me this will not work. I have several applications where I use a custom-made code to create the redis connection and we don't use a service provider. The major reason for this is because of many applications we migrated from very old dotnet versions.

I would like to enable the instrumentation statically anywhere in my code and I really think this would be the best case scenario and work for all use cases...

**Comment by CodeBlanch** [comment]
@lucasoares

> and we don't use a service provider

So with OTel .NET SDK there is _always_ an `IServiceCollection` \ `IServiceProvider`! Either one shared with the host (services.AddOpenTelemetry() style) or the SDK creates its own (Sdk.Create* style).

I think if we want to add an extension point that is fine, but we should use the DI API(s) because they are more standard.

I took a stab at what it might look like: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/compare/main...CodeBlanch:stackexchangeredis-creation-api?expand=1

Using that design we can accomplish the example on this PR as...

```csharp
       StackExchangeRedisInstrumentation? redisInstrumentation = null;

       using var tracerProvider = Sdk.CreateTracerProviderBuilder()
            .AddRedisInstrumentation()
            .AddConsoleExporter()
            .ConfigureServices(services => services.AddRedisInstrumentationCreatedAction(i => redisInstrumentation = i))
            .Build();

        // Connect to the server.
        using var connection = ConnectionMultiplexer.Connect("localhost:6379");
        redisInstrumentation?.AddConnection(connection);
```

Now I know what you are thinking right now. That may seem more complex than what you already have on this PR. Maybe! But I'm not really that concerned with your "old" style of doing things (sorry 🤣). What I want to make sure is that **new** styles work really well. Old styles should also work, but can be ugly.

What the new design allows us to do...

```csharp
builder.Services.AddOpenTelemetry();
builder.Services.AddMyRedisLibrary();
```

With some implementation like...

```csharp
public static class RedisExtensions
{
    public static IServiceCollection AddMyRedisLibrary(this IServiceCollection services)
    {
        services.AddSingleton<IRedisConnectionManager, RedisConnectionManager>();
        services.AddRedisInstrumentationCreatedAction((sp, name, instrumentation) =>
        {
            var redisConnectionManager = (RedisConnectionManager)sp.GetRequiredService<IRedisConnectionManager>();
            redisConnectionManager.AddInstrumentation(name, instrumentation);
        });
        services.ConfigureOpenTelemetryTracerProvider(
            (sp, builder) => builder.AddRedisInstrumentation());

        return services;
    }
}

public interface IRedisConnectionManager
{
    IConnectionMultiplexer GetConnection(string name);
}

internal sealed class RedisConnectionManager : IRedisConnectionManager
{
    private readonly Dictionary<string, StackExchangeRedisInstrumentation> instrumentation = new();

    public IConnectionMultiplexer GetConnection(string name)
    {
        IConnectionMultiplexer connection = this.FindOrCreateConnectionForName(name);

        if (this.instrumentation.TryGetValue(name, out var instrumentation))
        {
            instrumentation.AddConnection(connection);
        }
        else if (this.instrumentation.TryGetValue(string.Empty, out instrumentation))
        {
            instrumentation.AddConnection(connection);
        }

        return connection;
    }

    public void AddInstrumentation(
        string name,
        StackExchangeRedisInstrumentation instrumentation)
    {
        this.instrumentation.Add(name, instrumentation);
    }

    private IConnectionMultiplexer FindOrCreateConnectionForName(string name)
    {
        throw new NotImplementedException();
    }
}
```

Maybe that example is more compelling? Let me know what you think 😄

**Comment by lucasoares** [comment]
@CodeBlanch I don't think it is complex at all. If can make it static I think this should be enough..

> I'm like with 1 month using C# and dotnet and I'm not aware of the 'ecosystem-way' of doing things haha

I'm helping a team to instrument their applications, which happened to be kind old so they couldn't use the `services.AddOpenTelemetry() style` but the other one. The only feature missing for them is to be able to instrument many StackExchange.Redis connections in the code they have, which are created on-demand later after startup.

**Comment by lachmatt** [comment]
@CodeBlanch 
I think your suggestion from https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1139#issuecomment-1516764953 enables the scenarios that my changes from this PR tried to enable. Would you be willing to create a new PR with your changes, so that this one can be closed?

**Pull request closed without merging**